### DISCUSSION TITLE
Testing: How to deploy optiSLang application on Github

### DISCUSSION THREAD

**Discussion by ekrja** [ansys/pyoptislang#6] [Q&A] [answered]
Hello everyone,

I would like to ask you a question regarding testing on Github. For unit testing of pyOptiSLang, it is necessary to start optiSLang server process on Github. However, I'm not sure how to deploy optiSLang application on Github. I have found, that pyMAPDL uses Docker for this task. However, I have no experience with that. Is there any manual for this task? I am particularly interested in how to deploy and set up the ANSYS license server on Github.

Regards,
Jan

**Comment by rfahlberg** [comment]
Hi @pyansys/developers, can anyone point us to the correct direction for this topic? Thanks!

**Reply by smedikon** [reply]
@akaszynski/ @germa89: Can you help with this?

**Reply by da1910** [reply]
At the moment we have no facility to spin up an Ansys License Server, we have discussed approaches for running licensed software several times in PyAnsys developer meetings, and the consensus was that the current approach we use for PyMAPDL and grantami-bomanalytics is OK but not ideal.

I'm sure Alex or German will be able to provide more information, but you could take a look at how PyMAPDL handles it as a starting point.

**Reply by ekrja** [reply]
Thank you for your answer. I have already taken a look at the pyMAPDL and I have noticed the following settings in the GitHub workflow:
https://github.com/pyansys/pymapdl/blob/e2e0aa4daea3511d27848051c3c21c7e55a8389c/.github/workflows/nightly-mapdl-check.yml#L53
However, I still need some further explanation.

**Reply by germa89** [reply, marked as answer]
Thank you @smedikon for the mention.

Hi @ekrja 

The procedure for testing in GitHub our products will depend on the product itself. If a license is required, the things get slightly more complicated. 
What are you testing? A python package which connects to a product (MAPDL style, where PyMAPDL connects to MAPDL), or it is the product itself?

In PyMAPDL, the library connects to a MAPDL instance. To run this MAPDL instance we use docker. Creating the docker image is a totally different process (Can OptiSlang run in linux? Can you install it and have maximum 2gb image?)
Basically you will have to create a recipe file like:
```docker
FROM debian:9 # Using a base image

RUN echo stuff  # This line is just an example
RUN apt-get update; apt-get install -y vim  # This line is just an example

COPY --from=tini /usr/local /usr/local #copying files from your pc to your image
COPY --from=gosu /usr/local /usr/local # more copies

EXPOSE 50052 #exposing a port
RUN optyslang -options --more_options
```
Useful reads (not comprehensive list):
- https://docs.docker.com/engine/reference/builder/
- https://www.stereolabs.com/docs/docker/creating-your-image/
- https://www.mirantis.com/blog/how-do-i-create-a-new-docker-image-for-my-application/

You can ping me by email for more info. We can arrange a meeting and give you some guidance.

Once you have your docker image is working, and deployed in a registry; we can setup the CICD to pull it and run it, with the lines you mentioned. PyMAPDL and PyFluent have everything in place, you can check their repos to learn the procedure.

One issue is to a valid license, you will need access to the secrets which give you access to the license. You can ping me for more info. 

Summarizing:

- Get a working docker image of the product.
- Get the secrets for access to the license server
- Get library installed and running when using CICD for unit testing.

**Reply by akaszynski** [reply]
Awesome directions @german. Thanks!

**Reply by ekrja** [reply]
Hi @germa89

Thank you for your great answer. 

I would like to test python package (PyOptiSLang) which connects to the licensed product (optiSLang). Therefore, it is similar to PyMAPDL. The optiSLang can run on linux. However, its installation size is above 2 GB (about 4 GB). Is this a big problem?

I will try to get into it and do as much as I can and then give you some feedback how it goes.

Thank you for your support.

**Reply by rfahlberg** [reply]
@germa89, thank you for the great answer. 
@ekrja, we should be able to squeeze optiSLang in an 2GB image. Worst case, we can just manually delete stuff in the installation which is not needed for testing.

**Reply by ekrja** [reply]
@rfahlberg I think so that this issue has a solution. However, at first, we should be sure that 2 GB is a strict limit.

**Reply by dnwillia-work** [reply]
The images can be more than 2GB.  Fluent images are >>10 GB.

**Reply by ekrja** [reply]
@dnwillia-work Seems promising. Thank you.

**Reply by germa89** [reply]
The 2gb is not a strict limit. It is a recommendation. But you can go with something higher. Sorry for the misunderstanding.

**Reply by ekrja** [reply]
@germa89 That is great. Thank you for the clarification.

**Comment by koubaa** [comment]
@ekrja

"deploy optiSLang application on Github"

and

"unit testing of pyOptiSLang"

sound like very different things.  If its the latter - its something we have experience with for pyMAPDL + other cases.

**Reply by ekrja** [reply]
Thank you for your help. I'm sorry, I probably didn't choose correct words. Basically, I would like to start optiSLang application during pyOptiSLang unit testing on GitHub. So I mean the latter.

**Comment by dnwillia-work** [comment]
First you need to build a docker image of optiSLang and publish it to the GitHub container repo, then you need to setup your CI/CD pipeline to pull the image and run unit tests.  You can see examples of how this is working in the PyMAPDL and PyFluent repos.  Licensing wise just talk to @akaszynski as suggested by Sandeep, there is a way to do it.

**Reply by ekrja** [reply]
Thank you for the helpful advice. I didn't know that the same approach is also in PyFluent repo. I will take a look at it.
